File: robot-impl/src/main/java/ru/avca/robot/factory/BinanceFactory.java
package ru.avca.robot.factory;

import com.binance.api.client.BinanceApiClientFactory;
import io.micronaut.context.annotation.Factory;
import io.micronaut.context.annotation.Value;
import ru.avca.robot.utils.RobotUtils;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * @author a.chermashentsev
 * Date: 30.03.2021
 **/
@Factory
public class BinanceFactory {

    @Singleton
    public BinanceApiClientFactory binanceApiClientFactory(
            @Value("${exchange.binance.api.key}") String apiKey,
            @Value("${exchange.binance.api.secret}") String apiSecret) {
        return BinanceApiClientFactory.newInstance(apiKey.trim(), apiSecret.trim());
    }

    @Singleton
    @Inject
    public RobotUtils robotUtils(BinanceApiClientFactory binanceApiClientFactory) {
        return new RobotUtils(binanceApiClientFactory);
    }
}

File: backtest/src/main/kotlin/ru/avca/backtest/factory/BinanceFactory.kt
package ru.avca.backtest.factory

import com.binance.api.client.BinanceApiAsyncRestClient
import com.binance.api.client.BinanceApiClientFactory
import com.binance.api.client.BinanceApiRestClient
import io.micronaut.context.annotation.Factory
import io.micronaut.context.annotation.Value
import javax.inject.Singleton

@Factory
class BinanceFactory {
    @Singleton
    fun binanceApiClientFactory(
        @Value("\${exchange.binance.api.key}") apiKey: String,
        @Value("\${exchange.binance.api.secret}") apiSecret: String
    ): BinanceApiClientFactory {
        return BinanceApiClientFactory.newInstance(apiKey.trim { it <= ' ' }, apiSecret.trim { it <= ' ' })
    }

    @Singleton
    fun binanceAsyncApi(binanceApiClientFactory: BinanceApiClientFactory): BinanceApiAsyncRestClient {
        return binanceApiClientFactory.newAsyncRestClient()
    }

    @Singleton
    fun binanceSyncApi(binanceApiClientFactory: BinanceApiClientFactory): BinanceApiRestClient {
        return binanceApiClientFactory.newRestClient();
    }
}